src/pybind/mgr/mgr_util.py

```diff
@@ -164,3 +166,65 @@ def verify_tls_files(cert_fname, pkey_fname):
         logger.warning(
             'Private key {} and certificate {} do not match up: {}'.format(
                 pkey_fname, cert_fname, str(e)))
+
+
+def get_rates_from_data(data, use_abs=None):
+    """ Rates from time series data
+
+    :param data: Time series data
+    :type data: list[tuple[int, float]]
+
+    :param use_abs: If the derivative returned should be an absolute one
+    :type use_abs: bool (default None)
+        Using the absolute value is a temporary workaround for the pg decrease issue, that will
+        reduce the maximum amount in all time series data of a pool. Normally the maximum amount
+        should always increase.
+
+    :return: The derivative between all time series data points
+    :rtype: list[tuple[int, float]]
+
+    >>> get_rates_from_data([])
+    []
+    >>> get_rates_from_data([[1, 42]])
+    []
+    >>> get_rates_from_data([[0, 0], [2, 1], [3, 0], [4, -2]])
+    [(2, 0.5), (3, -1.0), (4, -2.0)]
+    >>> get_rates_from_data([[0, 0], [2, 1], [3, 0], [4, -2]], True)
+    [(2, 0.5), (3, 1.0), (4, 2.0)]
+    """
+    if not data or len(data) == 1:
+        return []
+    return [(data2[0], derivative(data1, data2, use_abs)) for data1, data2 in pairwise(data)]
+
+
+def derivative(p1, p2, use_abs=False):
+    """ Derivative between two time series data points
+
+    :param p1: Time series data point
+    :type p1: tuple[int, float]
+    :param p2: Time series data point
+    :type p2: tuple[int, float]
+
+    :param use_abs: If the derivative returned should be an absolute one
+    :type use_abs: bool (default False)
+        Using the absolute value is a temporary workaround for the pg decrease issue, that will
+        reduce the maximum amount in all time series data of a pool. Normally the maximum amount
+        should always increase.
+
+    :return: Derivative between both points
+    >>> values = [0, 1]
+    >>> times = [0, 0]
+    >>> derivative(*zip(times, values))
+    1.0
+    >>> times = [0, 2]
+    >>> derivative(*zip(times, values))
+    0.5
+    >>> values = [1, 0]
+    >>> derivative(*zip(times, values))
+    -0.5
+    >>> derivative(*zip(times, values), True)
+    0.5
+    """
+    time_diff = float(p2[0] - p1[0])
+    deriv = (p2[1] - p1[1]) / (time_diff if time_diff != 0 else 1)
```

Comment:
I still don't understand why we assume 1 second/time-unit interval when `t0 == t1`?

If that ever happens, let's either drop that sample, so no rate at all, or trigger an Exception, so the outer caller knows this weird condition happened.

Reply:
We don't. The reason to use one here is purely mathematically as we can't divide through zero.
If the value still is the same nothing changed through the two samples.

Reply:
That's wrong, and I'm sorry to be that picky with this code, but when you implement a math operation you cannot be creative. Imagine that some creative Python core developer would have decided that, to save you from the pain of dealing with exceptions, when you divide by 0, instead of raising `ZeroDivisionError`, Python would do the division by 1. Python would be the most unreliable programming language ever.

Any sample with the same timestamp should be considered an invalid one, and we should trigger an exception as soon we have them (before spreading them into other methods).

Why? Here there are 2 possible inputs when `time_diff = 0` and these are the outputs according to this code:

- `p2[1] == p1[1]`: then `rate = 0`. 
  - That basically means a dashboard operator will see a notorious drop to zero in whatever graph they're watching at, and if they're serious about their jobs they'll immediately start an investigation to find why this is happening: network packet drops? Hard-disk read retries?

- `p2[1] != p1[1]`: then `rate = X`. 
  - How do you decide whether rate is `X` or `-X`, since both samples belong to the same timestamp? Is it an increasing or a decreasing rate?

Reply:
I see your point, and we already discussed that in the other PR that was closed, that it's possible for some unknown reason...

The position in the time series data would determine who was recorded first.

As the time series data contains multiple segments I would not want to raise an error because of a minor issue that's not as tragic. IMO if we can predict it and know how to work around it we should do that instead of raising errors for a bad user experience.

How about we just drop the segment?

After reading your comment this sounds like the best idea IMO, we could also log the double entry to make sure it makes some noise even if it's little.

Reply:
> As the time series data contains multiple segments I would not want to raise an error because of a minor issue that's not as tragic. IMO if we can predict it and know how to work around it we should do that instead of raising errors for a bad user experience.

Sorry I cannot agree with this: data can be valid or invalid. You cannot let some data pass because it's not _that invalid_.

- Duplicate samples is valid data.
  - Returning rate 0 for them is **invalid data**.
- Mismatching samples for the same timestamp is **invalid data**.
  - Inferring any rate from them is **invalid data**.